lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,241 @@
+SYSCTL_HEADER = f"""# This config file was produced by sysctl lib v{LIBAPI}.{LIBPATCH}
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+"""
+
+
+class Error(Exception):
+    """Base class of most errors raised by this library."""
+
+    def __repr__(self):
+        """Represent the Error."""
+        return "<{}.{} {}>".format(type(self).__module__, type(self).__name__, self.args)
+
+    @property
+    def name(self):
+        """Return a string representation of the model plus class."""
+        return "<{}.{}>".format(type(self).__module__, type(self).__name__)
+
+    @property
+    def message(self):
+        """Return the message passed as an argument."""
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class SysctlConfig(Mapping[str, int]):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: Optional[str] = None) -> None:
+        self.name = name
+        self._data = self._load_data()
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> int:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def name(self) -> str:
+        """Name used to create the lib file."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        if value is None:
+            self._name, *_ = os.getenv("JUJU_UNIT_NAME", "unknown").split("/")
+        else:
+            self._name = value
+
+    @property
+    def charm_filepath(self) -> Path:
+        """Name for resulting charm config file."""
+        return Path(f"{SYSCTL_DIRECTORY}/90-juju-{self.name}")
+
+    @property
+    def charm_config_exists(self) -> bool:
+        """Return whether the charm config file exists."""
+        return os.path.exists(self.charm_filepath)
+
+    @property
+    def merged_config_exists(self) -> bool:
+        """Return whether a merged config file exists."""
+        return os.path.exists(SYSCTL_FILENAME)
+
+    def update(self, config: Dict[str, dict]) -> None:
+        """Update sysctl config options with a desired set of config params."""
+        self._parse_config(config)
+
+        # NOTE: case where own charm calls update() more than once. Remove first so
+        # we don't get validation errors.
+        if self.charm_config_exists:
+            self.remove()
```

Comment:
After demo, I realized that we need to create snapshot before this file is removed, because if something failed we should be able to restore it and all configuration coming from it.

Reply:
Yes, I was thinking that a `_restore_file_from_snapshot()` could be a good solution here.

Reply:
@zmraul did we fix this?

Reply:
Yes, instead of `remove()`, calling `_merge()` with the extra parameter of merging with/without own file takes care of this.